pkg/api/handlers/compat/images.go

```diff
@@ -45,6 +45,12 @@ func mergeNameAndTagOrDigest(name, tagOrDigest string) string {
 		// We have a digest, so let's change the separator.
 		separator = "@"
 	}
+
+	// if both name and tagOrDigest are defined we try to use the tagOrDigest image part and override the tag/digest
+	if len(name) > 0 && len(tagOrDigest) > 0 && strings.Contains(name, separator) {
+		parts := strings.Split(name, separator)
+		return fmt.Sprintf("%s%s%s", strings.Join(parts[:len(parts)-1], ""), separator, tagOrDigest)
+	}
```

Comment:
checking `len(tagOrDigest) > 0` is unnecessary as the function returns early if that is empty so we know it cannot be empty here

Then using string.Contains() and strings.Split() adds more overhead then needed here. You can use
`strings.LastIndex(name, separator)` once and then when the index is > 0 you can trim the name with `name[:index]` 

However more important speaking string parsing registry references like this does not look correct to me. In particular this will fail when a registry host includes a port number, i.e. `127.0.0.1:5000/myimage` as this would cause the code to cut at the wrong place. I am not super familiar with c/image but assume we should have some "proper" parsing code there to deal with that already so maybe we could resuse that?
cc @mtrmac 

If there is no such code we can use then we must at least ensure the `:` is after the first `/`

Reply:
right, I'll fix it

Reply:
would url parsing work ?

Reply:
> If there is no such code we can use then we must at least ensure the : is after the first /

@Luap99 do you think this is actually needed? I believe that the last occurrence of `:` or `@` should be enough, but I am not that familiar with registry/image uri

Reply:
`127.0.0.1:5000/myimage` is a valid image reference (i.e. not tag default to latest). In this case you would join the tag at the wrong place for `name=127.0.0.1:5000/myimage` and `tagOrDigest=mytag` will result in a broken `127.0.0.1:mytag` AFAICS

I think it would be really good to add unit tests for this function to check these corner cases properly

Reply:
yup, to be honest, I would split this functionality into a separate function and unit test that

Reply:
I mean you could test the entire mergeNameAndTagOrDigest(), I don't think it needs to be split smaller than that but I would not be opposed either

Reply:
I really strongly want this to use the standard `c/image/docker/reference` parser and not custom code.

Reply:
E.g. consider that we’ve had a request for IPv6 addresses (with their embedded `:`s) for years, and for all I know, it might actually happen some day.